src/core/network/polling.ts

```diff
@@ -0,0 +1,86 @@
+import * as tt from '../../telegram-types'
+import AbortController from 'abort-controller'
+import ApiClient from './client'
+import d from 'debug'
+import { promisify } from 'util'
+import { TelegramError } from './error'
+const debug = d('telegraf:polling')
+const wait = promisify(setTimeout)
+function always<T>(x: T) {
+  return () => x
+}
+const noop = always(Promise.resolve())
+
+export class Polling {
+  private readonly abortController = new AbortController()
+  private offset = 0
+  constructor(
+    private readonly telegram: ApiClient,
+    private readonly allowedUpdates: readonly tt.UpdateType[]
+  ) {}
+
+  private async *[Symbol.asyncIterator]() {
+    debug('Starting long polling')
+    do {
+      try {
+        const updates = await this.telegram.callApi(
+          'getUpdates',
+          {
+            timeout: 50,
+            offset: this.offset,
+            allowed_updates: this.allowedUpdates,
+          },
+          this.abortController
+        )
+        const last = updates[updates.length - 1]
+        if (last !== undefined) {
+          this.offset = last.update_id + 1
+        }
+        yield updates
+      } catch (err) {
+        if (err.name === 'AbortError') return
+        if (
+          err.name === 'FetchError' ||
+          (err instanceof TelegramError && err.code >= 500)
+        ) {
+          const retryAfter: number = err.parameters?.retry_after ?? 5
+          debug('Failed to fetch updates, retrying after %ds.', retryAfter, err)
+          await wait(retryAfter * 1000)
+          continue
+        }
+        if (
+          err instanceof TelegramError &&
+          (err.code === 401 /* Unauthorized */ ||
+            err.code === 409) /* Conflict */
+        ) {
+          this.confirmUpdates = noop
+          throw err
+        }
+        throw err
+      }
+    } while (!this.abortController.signal.aborted)
+  }
+
+  private async confirmUpdates() {
+    debug('Syncing update offset...')
+    await this.telegram.callApi('getUpdates', { offset: this.offset, limit: 1 })
+  }
+
+  async loop(handleUpdates: (updates: tt.Update[]) => Promise<void>) {
+    try {
+      for await (const updates of this) {
+        await handleUpdates(updates)
+      }
+    } finally {
+      debug('Long polling stopped')
+      // prevent instance reuse
+      this.abortController.abort()
+      await this.confirmUpdates().catch(noop)
+    }
```

Comment:
Just use this.
```suggestion
    try {
      for await (const updates of this) {
        await handleUpdates(updates)
      }
    } catch (err) {
      if (
        !(
          err instanceof TelegramError &&
          (err.code === 401 || err.code === 409)
        )
      ) {
        await this.confirmUpdates().catch(noop)
      }
    } finally {
      debug('Long polling stopped')
    }
```
Now you can get rid of
1) some error branching logic above
2) that unconventional method rewriting stuff ([this mistake was made before](https://github.com/telegraf/telegraf/blob/359f4f106dddccabfa6523c3bbd6ddf79a8b826a/extra.js#L45-L46))

Also, there's in fact no need for util flags and its perfectly readable.

Reply:
Pretty good, but slightly wrong: `confirmUpdates` is not called on successful graceful stop.

Reply:
> I can abort the controller in `finally`

That might still be a good idea if you still want to ensure that `Polling` objects won't be reused. Note that I skipped those lines in the suggestion above.

Reply:
I'm missing the rethrowing part! We have to rethrow the error in the new catch block of course. Don't use the suggestion as-is :smile:

Reply:
> Pretty good, but slightly wrong: confirmUpdates is not called on successful graceful stop.

Only saw that now, you're right, fixed it

Reply:
If the `catch` this is not DRY enough, you can move it to `confirmUpdates` internally.

Reply:
How? it has no access to `err`.

Reply:
And yes, your suggestion is not DRY enough, it's not clearly superior to mine.

Reply:
If you don't have anything better, why don't you approve this, and perhaps change it next time you'll be touching the file?

Reply:
Retrying logic could be extracted into a separate function and reused for `getMe` in `.launch()` and `.handleUpdate()`, or even exported to library users; then it could no longer concern itself with internals of `Polling`.

Reply:
> And yes, your suggestion is not DRY enough, it's not clearly superior to mine.

You do not understand what I was going for. However, given that you just merged before we could reach consensus, I'll clean up the mess in another PR.